purchase_delivery_split_date/models/purchase.py

```diff
@@ -1,138 +1,109 @@
 # Copyright 2014-2016 Numérigraphe SARL
 # Copyright 2017 ForgeFlow, S.L.
+# Copyright 2021 Jacques-Etienne Baudoux (BCIM) <je@bcim.be>
 # License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl.html).
 
-from itertools import groupby
-
 from odoo import api, fields, models
 
 
 class PurchaseOrderLine(models.Model):
     _inherit = "purchase.order.line"
 
-    @api.model
-    def _get_group_keys(self, order, line, picking=False):
-        """Define the key that will be used to group. The key should be
-        defined as a tuple of dictionaries, with each element containing a
-        dictionary element with the field that you want to group by. This
-        method is designed for extensibility, so that other modules can add
-        additional keys or replace them by others."""
-        date = fields.Date.context_today(self.env.user, line.date_planned)
-        # Split date value to obtain only the attributes year, month and day
-        key = ({"date_planned": fields.Date.to_string(date)},)
-        return key
-
-    @api.model
-    def _first_picking_copy_vals(self, key, lines):
-        """The data to be copied to new pickings is updated with data from the
-        grouping key.  This method is designed for extensibility, so that
-        other modules can store more data based on new keys."""
-        vals = {"move_lines": []}
-        for key_element in key:
-            if "date_planned" in key_element.keys():
-                vals["scheduled_date"] = key_element["date_planned"]
-        return vals
+    def _is_valid_picking(self, picking):
+        date_planned = picking.dt2date(self.date_planned)
+        expected_dates = {
+            picking.dt2date(expected_date)
+            for expected_date in picking.move_lines.mapped("date_expected")
+        }
+        return (
+            picking.state not in ("done", "cancel")
+            and picking.dt2date(picking.scheduled_date) == date_planned
+            and len(expected_dates) == 1
+            and picking.location_dest_id.usage in ("internal", "transit", "customer")
+        )
 
     def _create_stock_moves(self, picking):
-        """Group the receptions in one picking per group key"""
+        # _prepare_stock_moves needs previous move to be created to ensure the
+        # new move is inserted in the right picking. So create move one by one.
         moves = self.env["stock.move"]
-        # Group the order lines by group key
-        order_lines = sorted(
-            self.filtered(lambda l: not l.display_type), key=lambda l: l.date_planned
-        )
-        date_groups = groupby(
-            order_lines, lambda l: self._get_group_keys(l.order_id, l, picking=picking)
-        )
+        for line in self:
+            moves |= super(PurchaseOrderLine, line)._create_stock_moves(picking)
```

Comment:
I expect += to be a little faster here: these are new records so there should not be any duplicates.

Reply:
changed